Why can't we? (#13)
JelleZijlstra/JelleZijlstra.github.io: why-cant-we.md

or type alias.

_TypedDict keyword arguments_: The class syntax for `TypedDict` currently
supports one keyword argument, `total=False`, and the draft
[PEP 728](https://peps.python.org/pep-0728/) proposes adding more. It is not
obvious how these keyword arguments would be represented in the inline syntax.

_Generics_: If a generic inline TypedDict is defined using the old `TypeAlias`
syntax, like `MyDict: TypeAlias = {"a": T}`, it is impossible to specialize
the generic at runtime; `MyDict[int]` would throw KeyError. However, the native
syntax for type aliases in Python 3.12 fixes this; you could write
`type MyDict[T] = {"a": T}` and `MyDict[int]` would work as expected, because
subscripting gets handled by the type alias object, not the dict. Still, this could be
a problem for users who want to use the inline syntax and maintain compatibility
with Python 3.11 and earlier.

## Tuples as tuple types

### Idea

Allow writing `(int, str)` as a type instead of `tuple[int, str]`.
A single-element tuple type would be written as `(int,)`. Variadic
tuple types (`tuple[int, ...]`) could potentially be written as
`(int, ...)`.

> ### Problems

Reviewer: This also runs into problems with the `|` operator, similar to using `{}` for `TypedDict`s. We could add `tuple.__or__`, but that would make no sense at runtime outside of type annotations; it would be really baffling if we had this behaviour:

```pycon
>>> (int, str) | (int, bytes)
(int, str) | (int, bytes)
```

or worse:

```pycon
>>> (int, str) | (int, bytes)
tuple[int, str] | tuple[int, bytes]
```

Author: Thanks, good point! I'll add this and also add the "bare literals".

instead of concrete, invariant ones like `list` or `dict`. Providing overly
native-feeling syntax for list, set, and dict types goes against this advice.
The most convenient option should be the best one.

## Builtin functions as type expressions

### Idea

Three common typing-related objects have similar names to existing
builtin functions. What if we allowed users to use the builtin functions
in type expressions?

```python
x: any  # instead of typing.Any
y: iter[str]  # instead of collections.abc.Iterable[str]
z: callable[[str], int]  # instead of collections.abc.Callable[[str], int]
```

### Problems

_Muddying the waters_: Functions are not valid in type expressions. Special-casing
a few builtin functions for this would be confusing for users.

_Implementation complexity_: For this idea to work, the `iter` and `callable`
builtins would have to be made subscriptable. There is an open proposal
to make all functions subscriptable ([PEP 718](https://peps.python.org/pep-0718/)),
but failing that, we might have to add a new kind of builtin function just
for these two. Not out of the question, but it makes the language core more

> complicated.

Reviewer: Arguably this might actually conflict with PEP-718. PEP-718 would make it so that the meaning of a subscripted function would be to specialise the function. But under this idea, subscripting `callable` would not specialise the `callable` function (which is not generic)

Author: I did mean to say it would conflict, let me rewrite this for clarity.

```

Or in a type alias:

```python
type Movie = {"name": str, "year": int}
```

As a bonus, this also allows using keys that are not valid identifiers,
like the existing call-based syntax for creating a TypedDict.

The `NotRequired` and `ReadOnly` type qualifiers could also be used,
e.g. `{"name": str, "year": NotRequired[int]}`.

Note that some of the below problems can be mitigated by wrapping
the dictionary, e.g. `TypedDict[{"name": str}]`. The in-preparation
PEP 764 will propose this syntax.

### Problems

_`|` operator_: The `|` operator is already defined for dictionaries,
and its behavior doesn't match the expected behavior for unions in
the type system:

```pycon
>>> {"a": int} | {"b": str}
{'a': <class 'int'>, 'b': <class 'str'>}
>>> {"a": int} | {"a": str}
{'a': <class 'str'>}
```

This would make it difficult to introspect annotations using unions
of inline TypedDicts at runtime.

Creating a union of an inline TypedDict and a type object currently
throws an error:

```pycon
>>> {"a": int} | int
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unsupported operand type(s) for |: 'dict' and 'type'
```

> However, this is fixable by changing the implementation of `__or__`
> on type objects.

Reviewer: a bit nitpicky, but the specific example you give above this paragraph would actually have to be solved with a change to `type.__ror__` rather than `type.__or__`:

```suggestion
However, this is fixable by changing the implementation of `__or__`
and `__ror__` on type objects.
```

Author: To be even more nitpicky, those are actually the same :) They both map to the `nb_or` slot.

suggests using abstract, covariant types, such as `Sequence` or `Mapping`,
instead of concrete, invariant ones like `list` or `dict`. Providing overly
native-feeling syntax for list, set, and dict types goes against this advice.
The most convenient option should be the best one.

## Builtin functions as type expressions

### Idea

Three common typing-related objects have similar names to existing
builtin functions. What if we allowed users to use the builtin functions
in type expressions?

```python
x: any  # instead of typing.Any
y: iter[str]  # instead of collections.abc.Iterable[str]
z: callable[[str], int]  # instead of collections.abc.Callable[[str], int]
```

### Problems

_Muddying the waters_: Functions are not valid in type expressions. Special-casing
a few builtin functions for this would be confusing for users.

_Implementation complexity_: For this idea to work, the `iter` and `callable`
builtins would have to be made subscriptable. There is an open proposal
to make all functions subscriptable ([PEP 718](https://peps.python.org/pep-0718/)).
If that is accepted, subscripting would work, though there might be room for

> confusion because its meaning is different for these functions and for

Reviewer: ```suggestion
confusion because its meaning would be different for these functions than for
```
